<a href="https://colab.research.google.com/github/Alenushka2013/ML_for_people_tasks/blob/main/hw_3_3_Advanced_data_processing_methods_in_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('drive/MyDrive/DataLove_ML/Data/supermarket_sales.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [2]:
df.sort_values(by = 'Unit price', ascending = False).head(5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [3]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by=["Quantity", "Unit price"], ascending=[False, False]).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [4]:
def rating(x):
  if x >=8:
    return 'High'
  elif x > 6:
    return 'Medium'
  else:
    return 'Low'

df['Rating Category'] = df.Rating.apply(rating)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [5]:
df['Rating Category'].value_counts(normalize = True) * 100

Rating Category
Low       34.1
Medium    33.0
High      32.9
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [6]:
time = '11:28'

In [7]:
def get_hour(time):
  return int(time.split(':')[0])

get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [8]:
df['Hour'] = df['Time'].apply(get_hour)

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [9]:
df.groupby('Hour')['Invoice ID'].count()

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: Invoice ID, dtype: int64

Найбільше покупок здійснюють:
- о 10 годині - час відкриття магазину, ранні пташки
- о 13 годині - обідня перева
- 0 15 годині - закінчення уроків у школірів, студентів
- найбільше о 19 годині - закінчення робочого дня.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [10]:
df.groupby('City').Total.mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [11]:
df.groupby('City').agg({'Total': ['mean', 'min', 'max']})

Total                  
                 mean      min      max
City                                   
Mandalay   319.872506  18.6375  1022.49
Naypyitaw  337.099715  10.6785  1042.65
Yangon     312.354031  12.6945  1039.29

**Висновки:**

Найвищий середній чек — у Naypyitaw (337.10), при найнижчому мінімальному чеку може вказувати на вищу купівельну спроможність або ефективнішу роботу магазину.

Найнижчий середній чек — у Yangon (312.35), і другий за значнням мінімальний чек  може свідчити про менші покупки або знижений попит.

Максимальні чеки у Naypyitaw	і Yangon схожі — ≈1040, отже, максимальні покупки потенційно однакові, але середній чек у Naypyitaw вищий.

**Рекомендації для мережі супермаркетів:**

Розширити бізнес у Naypyitaw — місто має найвищий середній чек, що вказує на вигідний ринок.

Проаналізувати причини нижчого середнього чека в Yangon і Mandalay — можливо, потрібні локальні акції або зміни в асортименті.

Аналізувати товари, що купуються в Naypyitaw — можливо, саме певні категорії забезпечують високий дохід, і їх можна просувати в інших містах.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [12]:
df.to_csv('supermarket_sales_enriched.csv', index=False)